jax/experimental/pjit.py

```diff
@@ -95,29 +97,48 @@ def wrapped(*args, **kwargs):
     else:
       donated_invars = (False,) * len(args_flat)
 
-    in_axis_resources_flat = tuple(flatten_axes("pjit in_axis_resources",
-                                                in_tree, in_axis_resources))
-    out_axis_resources_thunk = HashableFunction(
-        lambda: tuple(flatten_axes("pjit out_axis_resources", out_tree(),
-                                   out_axis_resources)),
-        closure=(out_axis_resources_entries, out_axis_treedef))
-    _check_shapes_against_resources("pjit arguments", resource_env,
-                                    args_flat, in_axis_resources_flat)
-    flat_fun = _check_output_shapes(flat_fun, resource_env,
-                                    out_axis_resources_thunk)
-
-    out = pjit_call_p.bind(
-        flat_fun,
+    local_in_avals = tuple(core.raise_to_shaped(core.get_aval(a)) for a in args_flat)
+    jaxpr, in_axis_resources_flat, out_axis_resources_flat = \
+        _pjit_jaxpr(flat_fun, mesh, local_in_avals,
+                    in_tree, in_axis_resources,
+                    HashableFunction(out_tree, closure=()), out_axis_resources)
+
+    out = pjit_p.bind(
         *args_flat,
+        jaxpr=jaxpr,
         in_axis_resources=in_axis_resources_flat,
-        out_axis_resources_thunk=out_axis_resources_thunk,
+        out_axis_resources=out_axis_resources_flat,
         resource_env=resource_env,
         donated_invars=donated_invars,
         name=flat_fun.__name__)
     return tree_unflatten(out_tree(), out)
 
   return wrapped
 
+class _ListWithW(list):
+  __slots__ = ('__weakref__',)
+
+@lu.cache
```

Comment:
Use a `jax._src.util.cache` here? Then there's no weakrefs (so no need for `_ListWithW`), it's faster, and has bounded size with LRU logic. See the caches in lax/control_flow.py.

Reply:
Is that even possible? Are `lu.WrappedFun`s hashable? Why do we even use `lu.cache` in other places then? Note that this _is a final-style transform_.

Reply:
Maybe I need to look more closely at the context, but all other initial style primitives (i.e. in control_flow.py) do not need lu.cache. So why should pjit need it?

My understanding of this function was that it's not doing any transformation; rather it's just forming a jaxpr from `fun`, analogous to `_initial_style_jaxpr` in control_flow.py. Is there a difference I'm missing?

Reply:
I've tried changing it to use `@cache()`, but it turns out that it's not that simple, mostly due to `static_argnums` handling being written in an entirely `lu.WrappedFun`-centric way. I didn't feel like rewriting it all in this PR, so let's do this as a follow-up.